Documentation for writing functional tests and unit tests

The issue #395 made me think, that we lack documentation for:
- writing functional tests
- writing unit tests
- run tests before committing code

The only thing we have now is in CONTRIBUTING.md
https://github.com/webcompat/webcompat.com/blob/master/CONTRIBUTING.md#running-tests

which is a bit dry.

> You can run tests from the project root with the nosetests command.

For example, right now we get:

```
→ nosetests
....F...
======================================================================
FAIL: Test that the /issues route 307s to /index (for now).
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karl/code/webcompat.com/tests/test_urls.py", line 79, in test_issues_redirect
    self.assertEqual(rv.status_code, 307)
AssertionError: 200 != 307

----------------------------------------------------------------------
Ran 8 tests in 0.808s

FAILED (failures=1)
```

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -23,6 +23,9 @@ You are welcome to contribute to this project. Here are the guidelines we try to
 * [Coding](#coding)
 * [Running Tests](#running-tests)
   * [Functional Tests](#functional-tests)
+* [Writing Tests](#writing-tests)
+  * [Python Unit Tests](#python-unit-tests)
+  * [JS Functional Tests](#js-functional-tests)
 * [Production Server Setup](#production-server-setup)
 * [Acknowledgements](#acknowledgements)
 
@@ -336,7 +339,31 @@ node_modules/.bin/intern-runner config=tests/intern user=testusername pw=testpas
 
 **Note** Be aware that this will add the `testusername` and `testpassword` to your bash history. It is possible to run the tests without using a GitHub username and password as command-line arguments. In that case, the automatic login will fail and you then have 10 seconds to manually enter a username and password in the GitHub login screen that appears.
 
-**Note**: It's possible to run the tests without using a GitHub username and password as command-line arguments. In that case, the automatic login will fail and you then have 10 seconds to manually enter a username and password in the GitHub login screen that appears.
+## Writing Tests
+
+Contributions that add or modify major functionality to the project should typically come with tests to ensure we're not breaking things (or won't in the future!). There's always room for more testing, so general contributions in this form are always welcome.
+
+### Python Unit Tests
+
+Our Python unit tests are vanilla flavored [`unittest`](https://docs.python.org/2/library/unittest.html) tests. Unit tests placed in the `tests` directory will be automatically detected by nose&mdash;no manual registration is necessary.
+
+Unit tests are preferred for features or functionality that are independent of the browser front-end, i.e., API responses, application routes, etc.
+
+Important documentation links: 
+* [Writing nose tests](https://nose.readthedocs.org/en/latest/writing_tests.html)
+* [`unittest`](https://docs.python.org/2/library/unittest.html)
+* [Testing Flask](http://flask.pocoo.org/docs/0.10/testing/)
+
+### JS Functional Tests
+
+Functional tests are written in JavaScript, using [Intern](http://theintern.io/). There's a nice [guide on the Intern wiki](https://github.com/theintern/intern/wiki/Writing-Tests-with-Intern#functional-testing) that should explain enough to get you started. 
+
+Important documentation links:
+* [Leadfoot](https://theintern.github.io/leadfoot/): the library that drives the browser (via Selenium).
+* [ChaiJS](http://chaijs.com/api/assert/): the library used for assertions. 
+* [Intern wiki](https://github.com/theintern/intern/wiki): contains useful examples. 
+
+It's also recommended to look at the other test files in the `tests/functional` directory to see how things are commonly done.
 
 ## Production Server Setup
 

